File: src/epx_t2d.erl
%%%---- BEGIN COPYRIGHT --------------------------------------------------------
%%%
%%% Copyright (C) 2007 - 2012, Rogvall Invest AB, <tony@rogvall.se>
%%%
%%% This software is licensed as described in the file COPYRIGHT, which
%%% you should have received as part of this distribution. The terms
%%% are also available at http://www.rogvall.se/docs/copyright.txt.
%%%
%%% You may opt to use, copy, modify, merge, publish, distribute and/or sell
%%% copies of the Software, and permit persons to whom the Software is
%%% furnished to do so, under the terms of the COPYRIGHT file.
%%%
%%% This software is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY
%%% KIND, either express or implied.
%%%
%%%---- END COPYRIGHT ----------------------------------------------------------
%%% @author Tony Rogvall <tony@rogvall.se>
%%% @doc
%%%   2D transformations
%%% @end
%%% Created : 16 Oct 2011 by Tony Rogvall <tony@rogvall.se>

-module(epx_t2d).

-export([identity/0, translate/3, scale/3, rotate/2, compose/2]).
-export([deg_norm/1, deg_to_rad/1]).
-export([x/3, y/3, dx/3, dy/3, height/2, width/2, 
	 dimension/3, dimension/2, delta/3, delta/2,
	 point/3, point/2, points/2, 
	 rectangle/5, rectangle/2, rectangle_to_poly/2
	]).

-include("../include/epx_t2d.hrl").

-type(t2d() :: #t2d{}).
-export_type([t2d/0]).

-define(xp(T,X,Y), (X)*(T)#t2d.sx + (Y)*(T)#t2d.ry + (T)#t2d.tx).
-define(yp(T,X,Y), (X)*(T)#t2d.rx + (Y)*(T)#t2d.sy + (T)#t2d.ty).
-define(wp(T,W), abs((W)*(T)#t2d.sx)).
-define(hp(T,H), abs((H)*(T)#t2d.sy)).

dx(T, X, Y) -> X*T#t2d.sx + Y*T#t2d.ry.
dy(T, X, Y) -> X*T#t2d.rx + Y*T#t2d.sy.

x(T, X, Y) -> ?xp(T,X,Y).
y(T, X, Y) -> ?yp(T,X,Y).

height(T, H) -> ?hp(T,H).
width(T, W) ->  ?wp(T,W).

dimension(T, {W, H}) ->
    dimension(T, W, H).

dimension(T, W, H) ->
    {?hp(T,W), ?wp(T,H)}.


point(T, {X, Y}) ->
    point(T, X, Y).
point(T, X, Y) ->
    {?xp(T,X,Y), ?yp(T,X,Y)}.

points(T, XYs) ->
    #t2d { sx=Sx,ry=Ry,rx=Rx,sy=Sy,tx=Tx,ty=Ty } = T,
    [{ X*Sx + Y*Ry + Tx, X*Rx + Y*Sy + Ty } || {X,Y} <- XYs].
	      
delta(T, {Dx,Dy}) ->
    delta(T, Dx, Dy).

delta(T, Dx, Dy) ->
    {Dx*T#t2d.sx + Dy*T#t2d.ry,
     Dx*T#t2d.rx + Dy*T#t2d.sy }.

rectangle(T, {X, Y, W, H}) ->
    rectangle(T, X, Y, W, H).    

rectangle(T, X, Y, W, H) ->
    {?xp(T,X,Y), ?yp(T,X,Y), W*T#t2d.sx, H*T#t2d.sy}.

rectangle_to_poly(T, {X,Y,W,H}) ->
    points(T, [{X,Y},{X+W,Y},{X+W,Y+H},{X,Y+H}]).

%%
%% Get the identity transformation
%%

identity() ->
    #t2d {}.

%% @doc
%% Compose - Transform "matrix"
%%  | a b e | | sx  ry  tx |   | a*sx+b*rx  a*ry+b*sy a*tx+b*ty+e |
%%  | c d f |*| rx  sy  ty | = | c*sx+d*rx  c*ry+d*sy c*tx+d*ty+f |
%%  | 0 0 1 | | 0   0   1  |   | 0          0         1           |
%% @end
-spec compose(t2d(), t2d()) -> t2d().
compose(T, S) ->
    A = S#t2d.sx,
    B = S#t2d.ry,
    C = S#t2d.rx,
    D = S#t2d.sy,
    E = S#t2d.tx,
    F = S#t2d.ty,

    Sx = A*T#t2d.sx + B*T#t2d.rx,
    Ry = A*T#t2d.ry + B*T#t2d.sy,
    Tx = A*T#t2d.tx + B*T#t2d.ty + E,

    Rx = C*T#t2d.sx + D*T#t2d.rx,
    Sy = C*T#t2d.ry + D*T#t2d.sy,
    Ty = C*T#t2d.tx + D*T#t2d.ty + F,
    T#t2d { sx=Sx, rx=Rx, ry=Ry, sy=Sy, tx=Tx, ty=Ty}.

%% @doc
%% Translate - Transform "matrix"
%%  | sx  ry  tx |   | 1 0 x |     | sx ry sx*x+ry*y+tx |
%%  | rx  sy  ty | * | 0 1 y |  =  | rx sy rx*x+sy*y+ty |
%%  | 0   0   1  |   | 0 0 1 |     | 0  0  1            |
%% @end
-spec translate(t2d(), number(), number()) -> t2d().
translate(T,Tx,Ty) when is_number(Tx), is_number(Ty) ->
    Tx1 = T#t2d.sx*Tx + T#t2d.ry*Ty + T#t2d.tx,
    Ty1 = T#t2d.rx*Tx + T#t2d.sy*Ty + T#t2d.ty,
    T#t2d { tx=Tx1, ty=Ty1}.

%% @doc
%%  Scale - Transform "matrix"
%%  | sx  ry  tx |   | x   0   0 |    | sx*x ry*y tx |
%%  | rx  sy  ty | * | 0   y   0 | =  | rx*x sy*y ty |
%%  | 0   0   1  |   | 0   0   1 |    | 0    0    1  |
%% @end
-spec scale(t2d(), number(), number()) -> t2d().
scale(T, Sx, Sy) when is_number(Sx), is_number(Sy) ->
    Sx1 = T#t2d.sx*Sx,
    Ry1 = T#t2d.ry*Sy,
    Rx1 = T#t2d.rx*Sx,
    Sy1 = T#t2d.sy*Sy,
    T#t2d { sx=Sx1, ry=Ry1, rx=Rx1, sy=Sy1 }.

%% @doc
%% Rotate - Transform "matrix"
%%  | sx  ry  tx |   | c  -s   0 |     | sx*c+ry*s  -sx*s+ry*c tx |
%%  | rx  sy  ty | * | s   c   0 |  =  | rx*c+sy*s  -rx*s+sy*c ty |
%%  | 0   0   1  |   | 0   0   1 |     | 0          0          1  |
%% @end
-spec rotate(t2d(), number()) -> t2d().
rotate(T, Angle) ->
    A = deg_to_rad(deg_norm(Angle)),
    C = math:cos(A),
    S = math:sin(A),

    Sx =  T#t2d.sx*C + T#t2d.ry*S,  Ry = -T#t2d.sx*S + T#t2d.ry*C,
    Rx =  T#t2d.rx*C + T#t2d.sy*S,  Sy = -T#t2d.rx*S + T#t2d.sy*C,

    T#t2d { sx=Sx, rx=Rx, ry=Ry, sy=Sy }.

%% @doc
%% Normalize angle to -180..180
%% @end
deg_norm(A) when is_number(A) ->
    N = trunc(A / 360),
    if A < 0 ->
	    A - ((N-1)*360);
       true ->
	    A - N*360
    end.

%% @doc
%% Convert degrees to radians
%% @end

deg_to_rad(Deg) ->
    Deg * (math:pi()/180.0).

File: c_src/epx_t2d.c
/***************************************************************************
 *
 * Copyright (C) 2007 - 2021, Rogvall Invest AB, <tony@rogvall.se>
 *
 * This software is licensed as described in the file COPYRIGHT, which
 * you should have received as part of this distribution. The terms
 * are also available at http://www.rogvall.se/docs/copyright.txt.
 *
 * You may opt to use, copy, modify, merge, publish, distribute and/or sell
 * copies of the Software, and permit persons to whom the Software is
 * furnished to do so, under the terms of the COPYRIGHT file.
 *
 * This software is distributed on an "AS IS" basis, WITHOUT WARRANTY OF ANY
 * KIND, either express or implied.
 *
 ***************************************************************************/

#include "../include/epx_t2d.h"
#include <math.h>
#include <memory.h>

const epx_t2d_t epx_identity_ctm =
{
    .sx = 1.0f, .ry = 0.0f, .tx = 0.0f,
    .rx = 0.0f, .sy = 1.0f, .ty = 0.0f,
    .version = 0
};

void epx_t2d_transform_xy(epx_t2d_t* t, float* xp, float* yp)
{
    float x = *xp;
    float y = *yp;
    *xp = x*t->sx + y*t->ry + t->tx;
    *yp = x*t->rx + y*t->sy + t->ty;
}

void epx_t2d_transform_wh(epx_t2d_t* t, float* wp, float* hp)
{
    *wp = fabs((*wp)*t->sx);
    *hp = fabs((*hp)*t->sy);
}

void epx_t2d_transform_vec(epx_t2d_t* t, float* vec, size_t n)
{
    while(n--) {
	epx_t2d_transform_xy(t, vec, vec+1);
	vec += 2;
    }
}

void epx_t2d_identity(epx_t2d_t* dst)
{
    memcpy(dst, &epx_identity_ctm, sizeof(epx_t2d_t));
}

void epx_t2d_set(float elem[6], epx_t2d_t* dst)
{
    dst->version++;
    dst->sx = elem[T_SX];
    dst->ry = elem[T_RY];
    dst->tx = elem[T_TX];
    dst->ry = elem[T_RY];
    dst->sy = elem[T_SY];
    dst->ty = elem[T_TY];
}

void epx_t2d_get(epx_t2d_t* t, float elem[6])
{
    elem[T_SX] = t->sx;
    elem[T_RY] = t->ry;
    elem[T_TX] = t->tx;
    elem[T_RX] = t->rx;    
    elem[T_SY] = t->sy;
    elem[T_TY] = t->ty;
}

//
// Compose - Transform "matrix"
//  | sx  ry  tx |   | a b e |     | sx*a+ry*c  sx*b+ry*d sx*e+ry*f+tx |
//  | rx  sy  ty | * | c d f |  =  | rx*a+sy*c  rx*b+sy*d rx*e+sy*f+ty |
//  | 0   0   1  |   | 0 0 1 |     | 0          0         1            |
//
void epx_t2d_compose(epx_t2d_t* t, epx_t2d_t* s, epx_t2d_t* dst)
{
    float tsx = t->sx;
    float try = t->ry;
    float trx = t->rx;
    float tsy = t->sy;
    float sx = tsx*s->sx + try*s->rx;
    float ry = tsx*s->ry + try*s->sy;
    float rx = trx*s->sx + tsy*s->rx;
    float sy = trx*s->ry + tsy*s->sy;
    float tx = tsx*s->tx + try*s->ty + t->tx;
    float ty = trx*s->tx + tsy*s->ty + t->ty;
    dst->version++;
    dst->sx = sx;
    dst->ry = ry;
    dst->sy = sy;
    dst->rx = rx;
    dst->tx = tx;
    dst->ty = ty;
}

//
// Translate - Transform "matrix"
//  | sx  ry  tx |   | 1 0 x |     | sx ry sx*x+ry*y+tx |
//  | rx  sy  ty | * | 0 1 y |  =  | rx sy rx*x+sy*y+ty |
//  | 0   0   1  |   | 0 0 1 |     | 0  0  1            |
//
void epx_t2d_translate(epx_t2d_t* t, float tx, float ty, epx_t2d_t* dst)
{
    dst->version++;
    dst->tx = t->tx + t->sx*tx + t->ry*ty;
    dst->ty = t->ty + t->rx*tx + t->sy*ty;
    dst->sx = t->sx;
    dst->ry = t->ry;
    dst->sy = t->sy;
    dst->rx = t->rx;
}

//
// Scale - Transform "matrix"
//  | sx  ry  tx |   | x   0   0 |    | sx*x ry*y tx |
//  | rx  sy  ty | * | 0   y   0 | =  | rx*x sy*y ty |
//  | 0   0   1  |   | 0   0   1 |    | 0    0    1  |
//
void epx_t2d_scale(epx_t2d_t* t, float sx, float sy, epx_t2d_t* dst)
{
    dst->version++;    
    dst->sx = t->sx*sx;
    dst->ry = t->ry*sy;
    dst->rx = t->rx*sx;    
    dst->sy = t->sy*sy;
    dst->tx = t->tx;
    dst->ty = t->ty;
}

//
// Rotate - Transform "matrix"
//  | sx  ry  tx |   | c  -s   0 |     | sx*c+ry*s  -sx*s+ry*c tx |
//  | rx  sy  ty | * | s   c   0 |  =  | rx*c+sy*s  -rx*s+sy*c ty |
//  | 0   0   1  |   | 0   0   1 |     | 0          0          1  |
//

void epx_t2d_rotate(epx_t2d_t* t, float a, epx_t2d_t* dst)
{
    float c = cosf(a);
    float s = sinf(a);
    float ax, ay;
    
    ax =  t->sx*c + t->ry*s;
    ay = -t->sx*s + t->ry*c;

    dst->version++;
    dst->sx = ax;
    dst->ry = ay;
    dst->tx = t->tx;
    
    ax = t->rx*c + t->sy*s;
    ay = -t->rx*s + t->sy*c;

    dst->rx = ax;
    dst->sy = ay;
    dst->tx = t->tx;
}
